In [106]:
# Student Identity
LAST_NAME = "Laude"
STUDENT_ID = "TUPM-25-4764"

# Derivative computational parameters
seed_digit = int(STUDENT_ID[-1])
id_checksum = sum(int(d) for d in STUDENT_ID if d.isdigit())
vector_dim = len(LAST_NAME)

# Encapsulate state into a Dictionary (Associative Map)
sys_config = {
    "operator": LAST_NAME,
    "auth_id": STUDENT_ID,
    "base_seed": seed_digit,
    "checksum": id_checksum,
    "vector_dim": vector_dim,
    "status": "INITIALIZED",
}
# Display system state
print("=== SYSTEM CONFIGURATION ===")
for key, value in sys_config.items():
    print(f"{key.upper()}: {value}")

=== SYSTEM CONFIGURATION ===
OPERATOR: Laude
AUTH_ID: TUPM-25-4764
BASE_SEED: 4
CHECKSUM: 28
VECTOR_DIM: 5
STATUS: INITIALIZED


In [107]:
# Initialize a list using the system configuration
base_val = sys_config["base_seed"]
number_sequence = [base_val, base_val + 15, sys_config["checksum"]]

print(f"Initial Sequence: {number_sequence}")

# Append a new value to the sequence
number_sequence.append(base_val + 20)
print(f"After append: {number_sequence}")

Initial Sequence: [4, 19, 28]
After append: [4, 19, 28, 24]


In [108]:
new_numbers = [base_val + 5, sys_config["vector_dim"], base_val]
number_sequence.extend(new_numbers)
print(f"After Extend: {number_sequence}")

# Count occurrences of the base value
base_count = number_sequence.count(base_val)
print(f"Occurrences of {base_val}: {base_count}")

# Sort the sequence in ascending order
number_sequence.sort()
print(f"Sorted Sequence: {number_sequence}")

After Extend: [4, 19, 28, 24, 9, 5, 4]
Occurrences of 4: 2
Sorted Sequence: [4, 4, 5, 9, 19, 24, 28]


In [109]:
# Define an immutable tuple using system parameter
fixed_coordinates = (sys_config["vector_dim"], sys_config["base_seed"], 0)
print(f"Fixed Coordinates: {fixed_coordinates}")

# Unpack the tuple into individual variables
x_val, y_val, z_val = fixed_coordinates
print(f"Unpacked Values -> X:{x_val}, Y:{y_val}, Z:{z_val}")

# Attempt to modify the tuple (expected to fail)
print("\nAttempting modification...")
try:
    fixed_coordinates[0] = 99
except TypeError as error_msg:
    print(f"Modification Error: {error_msg}")

Fixed Coordinates: (5, 4, 0)
Unpacked Values -> X:5, Y:4, Z:0

Attempting modification...
Modification Error: 'tuple' object does not support item assignment


In [110]:
# Create a dictionary representing a structured data payload
data_payload = {
    "identifier": sys_config["auth_id"],
    "dimension": sys_config["vector_dim"],
    "status": "active",
}

# Iterate and display key-value pairs
print("--- Payload Data ---")
for key, value in data_payload.items():
    print(f"{key.capitalize()}: {value}")

# Add a new key-value pair dynamically
data_payload["efficiency_rating"] = 98.5
print(f"\nUpdated Payload: {data_payload}")

--- Payload Data ---
Identifier: TUPM-25-4764
Dimension: 5
Status: active

Updated Payload: {'identifier': 'TUPM-25-4764', 'dimension': 5, 'status': 'active', 'efficiency_rating': 98.5}


In [111]:
base_val = sys_config["base_seed"]

# Initialize a list with deliberate duplicate values
raw_data = [base_val, base_val + 5, base_val, sys_config["vector_dim"], base_val + 5]
print(f"Raw List (with duplicates): {raw_data}")

# Convert the list to a Set to remove duplicates
unique_data = set(raw_data)
print(f"Unique Set: {unique_data}")

Raw List (with duplicates): [4, 9, 4, 5, 9]
Unique Set: {9, 4, 5}


In [112]:
# Define a secondary reference set for comparison
reference_set = {base_val, base_val + 10, base_val + 20}
print(f"Reference Set: {reference_set}")

# Perform mathematical intersections and unions
common_elements = unique_data.intersection(reference_set)
combined_elements = unique_data.union(reference_set)

print(f"Intersection (Common): {common_elements}")
print(f"Union (Combined): {combined_elements}")

Reference Set: {24, 4, 14}
Intersection (Common): {4}
Union (Combined): {4, 5, 24, 9, 14}


In [113]:
import array

# Initialize an integer array using the type code 'i'
number_array = array.array(
    "i", [sys_config["base_seed"], sys_config["vector_dim"], 100]
)
print(f"Integer Array: {number_array}")

# Attempt to append a non-integer data type (Expected to fail)
print("\nAttempting to append a string...")
try:
    number_array.append("invalid")
except TypeError as error_msg:
    print(f"Type Error: {error_msg}")

Integer Array: array('i', [4, 5, 100])

Attempting to append a string...
Type Error: 'str' object cannot be interpreted as an integer


In [114]:
base_val = sys_config["base_seed"]

# Generate a sequence dynamically
generated_list = [(x * base_val) for x in range(1, 6)]
print(f"Generated Comprehension: {generated_list}")

# Filter the generated sequence
filtered_list = [x for x in generated_list if x > 15]
print(f"Filtered Comprehension (Values > 15): {filtered_list}")

Generated Comprehension: [4, 8, 12, 16, 20]
Filtered Comprehension (Values > 15): [16, 20]


In [115]:
from collections import deque

# Initialize a double-ended queue
data_queue = deque([sys_config["base_seed"], sys_config["vector_dim"]])
print(f"Initial Deque: {data_queue}")

# Append ti both ends
data_queue.append(100)
data_queue.appendleft(200)
print(f"After Appends: {data_queue}")

# Remove form the left side
data_queue.popleft()
print(f"Final Deque (After Popleft): {data_queue}")

Initial Deque: deque([4, 5])
After Appends: deque([200, 4, 5, 100])
Final Deque (After Popleft): deque([4, 5, 100])


In [116]:
from collections import defaultdict

# Initialize with a default integer factory (creates a 0 for missing keys)
default_data = defaultdict(int)

# Assign a know key
default_data["active_key"] = sys_config["vector_dim"]

print(f"Existing key Value: {default_data['active_key']}")

# Accessing an uninitialized ket will not throw a KeyError
print(f"Missing Key Value (Auto-generated): {default_data['Unknown_key']}")

Existing key Value: 5
Missing Key Value (Auto-generated): 0


### **Exercise 1**

In [117]:
# Exercise 1

STUDENT_MAJOR = "BSECE"  # Example degree program
FAVORITE_ARTIST = "RADIOHEAD"
LOCAL_HAZARD = "TYPHOON"  # Example local emergency event

# Derived computational parameters
SEED_NUM = int(STUDENT_ID[-1])  # Last digit: 4
CONTROL_NUM = max(3, SEED_NUM)  # CONTROL_NUM = 4

print(f"SEED_NUM: {SEED_NUM}")
print(f"STUDENT_MAJOR: {STUDENT_MAJOR}")
print(f"FAVORITE_ARTIST: {FAVORITE_ARTIST}")
print(f"LOCAL_HAZARD: {LOCAL_HAZARD}")
print(f"CONTROL_NUM: {CONTROL_NUM}")

SEED_NUM: 4
STUDENT_MAJOR: BSECE
FAVORITE_ARTIST: RADIOHEAD
LOCAL_HAZARD: TYPHOON
CONTROL_NUM: 4


In [118]:
# Import required collections
from collections import defaultdict, Counter

# Initialize simulated application stream
stream = [
    ("TUP Manila", "BSECE"),
    ("TUP Taguig", f"BSME_v{CONTROL_NUM}"),
    ("TUP Manila", f"BSME_v{CONTROL_NUM}"),
    ("TUP Manila", "BSECE"),
    ("TUP Visayas", STUDENT_MAJOR),
    ("TUP Taguig", "BSECE"),
    ("TUP Manila", "BSECE"),
]

for app in stream:
    print(app)
print(f"Initial stream length: {len(stream)}")

('TUP Manila', 'BSECE')
('TUP Taguig', 'BSME_v4')
('TUP Manila', 'BSME_v4')
('TUP Manila', 'BSECE')
('TUP Visayas', 'BSECE')
('TUP Taguig', 'BSECE')
('TUP Manila', 'BSECE')
Initial stream length: 7


In [119]:
# Data Expansion Phase

custom_applications = [
    ("TUP Manila", STUDENT_MAJOR),
    ("TUP Taguig", f"BSIT_v{CONTROL_NUM}"),
    ("TUP Visayas", FAVORITE_ARTIST),
    ("TUP Manila", LOCAL_HAZARD),
    ("TUP Manila", STUDENT_MAJOR),
    ("TUP Taguig", STUDENT_MAJOR),
    ("TUP Visayas", f"BSME_v{CONTROL_NUM}"),
]

# Extend the stream with custom applications
stream.extend(custom_applications)

print(f"Total stream length: {len(stream)}")
print("Custom applications added:")
for i, app in enumerate(custom_applications, 1):
    print(f"  {i}. {app}")

Total stream length: 14
Custom applications added:
  1. ('TUP Manila', 'BSECE')
  2. ('TUP Taguig', 'BSIT_v4')
  3. ('TUP Visayas', 'RADIOHEAD')
  4. ('TUP Manila', 'TYPHOON')
  5. ('TUP Manila', 'BSECE')
  6. ('TUP Taguig', 'BSECE')
  7. ('TUP Visayas', 'BSME_v4')


In [120]:
# Algorithm Phase 1: Aggregation

# Initialize defaultdict with list as default factory
campus_aggregator = defaultdict(list)

# Iterate through expanded stream and append programs to campus keys
for campus, program in stream:
    campus_aggregator[campus].append(program)

for campus, programs in campus_aggregator.items():
    print(f"\n{campus}:")
    print(f"  Total applications: {len(programs)}")
    print(f"  Programs: {programs}")


TUP Manila:
  Total applications: 7
  Programs: ['BSECE', 'BSME_v4', 'BSECE', 'BSECE', 'BSECE', 'TYPHOON', 'BSECE']

TUP Taguig:
  Total applications: 4
  Programs: ['BSME_v4', 'BSECE', 'BSIT_v4', 'BSECE']

TUP Visayas:
  Total applications: 3
  Programs: ['BSECE', 'RADIOHEAD', 'BSME_v4']


In [121]:
# Algorithm Phase 2: Frequency Analysis

# Analyze each campus using Counter
frequency_reports = {}

for campus, programs in campus_aggregator.items():
    # Create Counter for this campus
    program_counter = Counter(programs)
    frequency_reports[campus] = program_counter

    print(f"\n{campus} Program Frequencies:")
    # Display all programs with their frequencies
    for program, count in program_counter.most_common():
        print(f"  {program}: {count} application(s)")

# Special analysis for TUP Manila (top requested program)
manila_counter = frequency_reports["TUP Manila"]
top_manila_program = manila_counter.most_common(1)[0]

print("\nTUP MANILA TOP PROGRAM ANALYSIS")
print(f"  Top Requested Program: {top_manila_program[0]}")
print(f"  Frequency: {top_manila_program[1]} applications")


TUP Manila Program Frequencies:
  BSECE: 5 application(s)
  BSME_v4: 1 application(s)
  TYPHOON: 1 application(s)

TUP Taguig Program Frequencies:
  BSECE: 2 application(s)
  BSME_v4: 1 application(s)
  BSIT_v4: 1 application(s)

TUP Visayas Program Frequencies:
  BSECE: 1 application(s)
  RADIOHEAD: 1 application(s)
  BSME_v4: 1 application(s)

TUP MANILA TOP PROGRAM ANALYSIS
  Top Requested Program: BSECE
  Frequency: 5 applications


| Field | Student Entry |
|-------|---------------|
| CONTROL_NUM Used | 4 |
| STUDENT_MAJOR Used | BSECE |
| Expanded Stream Length (After adding data) | 14 |
| Total Applications Processed for TUP Manila | 7 |
| Top Requested Program at TUP Manila | BSECE |
| Exact Frequency of Top Program | 3 |

### **Exercise 2**

In [127]:
# Initialize Festival Lineup as a Set
festival = {
    "Ben&Ben",
    "SB19",
    "Bini",
    "Eraserheads",
    FAVORITE_ARTIST,
    "Zild",
    f"Indie Artist {CONTROL_NUM}",
}

# Initialize User Profiles as Sets
user_a = {"Ben&Ben", "Bini", "Maki", "Dionela", FAVORITE_ARTIST}  # BENANDBEN

user_b = {
    "SB19",
    "Eraserheads",
    "Zild",
    f"Indie Artist {CONTROL_NUM}",  # Indie Artist 4
    "Parokya ni Edgar",
}

print("INITIAL FESTIVAL LINEUP")
print(f"  Festival artists: {festival}")
print(f"  Total artists: {len(festival)}")

print("\nUSER PROFILES")
print(f"  User A: {user_a}")
print(f"  User B: {user_b}")

INITIAL FESTIVAL LINEUP
  Festival artists: {'Indie Artist 4', 'Eraserheads', 'SB19', 'Bini', 'Zild', 'Ben&Ben', 'RADIOHEAD'}
  Total artists: 7

USER PROFILES
  User A: {'Dionela', 'Bini', 'Maki', 'Ben&Ben', 'RADIOHEAD'}
  User B: {'Indie Artist 4', 'Eraserheads', 'SB19', 'Parokya ni Edgar', 'Zild'}


In [128]:
# Data Expansion Phase

new_artists = {"December Avenue", "Moira Dela Torre", "IVOS", "Munimuni"}

# Using .update() method to add multiple artists at once
festival.update(new_artists)

print(f"Added {CONTROL_NUM} new artists to festival:")
for artist in new_artists:
    print(f"  • {artist}")
print(f"\nExpanded Festival Lineup: {festival}")
print(f"Total artists after expansion: {len(festival)}")

Added 4 new artists to festival:
  • IVOS
  • Munimuni
  • Moira Dela Torre
  • December Avenue

Expanded Festival Lineup: {'IVOS', 'Munimuni', 'Indie Artist 4', 'Eraserheads', 'SB19', 'Bini', 'Zild', 'Moira Dela Torre', 'December Avenue', 'Ben&Ben', 'RADIOHEAD'}
Total artists after expansion: 11


In [129]:
# Jaccard Similarity Algorithm


def calculate_jaccard_similarity(set1, set2, user_name):
    """
    Calculate Jaccard Similarity between two sets
    J(A,B) = |A ∩ B| / |A ∪ B|
    Returns percentage value
    """
    # Calculate intersection and union
    intersection = set1.intersection(set2)
    union = set1.union(set2)

    # Calculate Jaccard Index
    if len(union) == 0:  # Avoid division by zero
        similarity = 0
    else:
        similarity = (len(intersection) / len(union)) * 100

    print(f"\n{user_name} Jaccard Analysis")
    print(f"  Intersection (artists in common): {intersection}")
    print(f"  Size of Intersection: {len(intersection)}")
    print(f"  Union (all unique artists): {union}")
    print(f"  Size of Union: {len(union)}")
    print(f"  Jaccard Similarity: {similarity:.2f}%")

    return similarity, intersection, union


# Calculate for User A
jaccard_a, intersect_a, union_a = calculate_jaccard_similarity(
    user_a, festival, "User A"
)

# Calculate for User B
jaccard_b, intersect_b, union_b = calculate_jaccard_similarity(
    user_b, festival, "User B"
)


User A Jaccard Analysis
  Intersection (artists in common): {'Ben&Ben', 'Bini', 'RADIOHEAD'}
  Size of Intersection: 3
  Union (all unique artists): {'Dionela', 'Munimuni', 'Indie Artist 4', 'Bini', 'Moira Dela Torre', 'December Avenue', 'Ben&Ben', 'RADIOHEAD', 'IVOS', 'Eraserheads', 'SB19', 'Zild', 'Maki'}
  Size of Union: 13
  Jaccard Similarity: 23.08%

User B Jaccard Analysis
  Intersection (artists in common): {'Indie Artist 4', 'Eraserheads', 'SB19', 'Zild'}
  Size of Intersection: 4
  Union (all unique artists): {'Munimuni', 'Indie Artist 4', 'Bini', 'Parokya ni Edgar', 'Moira Dela Torre', 'December Avenue', 'Ben&Ben', 'RADIOHEAD', 'IVOS', 'Eraserheads', 'SB19', 'Zild'}
  Size of Union: 12
  Jaccard Similarity: 33.33%


In [130]:
print("DEALBREAKERS ANALYSIS")

# Find User A's dealbreakers (artists they listen to that are NOT in festival)
dealbreakers_a = user_a.difference(festival)

print(f"\nUser A's Listening Profile: {user_a}")
print(f"Expanded Festival Lineup: {festival}")
print(f"\nUser A's DEALBREAKERS (artists NOT performing at festival):")
if dealbreakers_a:
    for artist in sorted(dealbreakers_a):
        print(f"  ✗ {artist}")
    print(f"\nTotal dealbreakers: {len(dealbreakers_a)}")
    print(
        f"Dealbreaker percentage: {(len(dealbreakers_a)/len(user_a))*100:.2f}% of User A's library"
    )
else:
    print("  ✓ None! All of User A's artists are performing at the festival.")

DEALBREAKERS ANALYSIS

User A's Listening Profile: {'Dionela', 'Bini', 'Maki', 'Ben&Ben', 'RADIOHEAD'}
Expanded Festival Lineup: {'IVOS', 'Munimuni', 'Indie Artist 4', 'Eraserheads', 'SB19', 'Bini', 'Zild', 'Moira Dela Torre', 'December Avenue', 'Ben&Ben', 'RADIOHEAD'}

User A's DEALBREAKERS (artists NOT performing at festival):
  ✗ Dionela
  ✗ Maki

Total dealbreakers: 2
Dealbreaker percentage: 40.00% of User A's library


In [131]:
print("FESTIVAL RECOMMENDATION ANALYSIS")

# Determine which user is a better match for the festival
if jaccard_a > jaccard_b:
    better_match = "User A"
    better_score = jaccard_a
    worse_score = jaccard_b
elif jaccard_b > jaccard_a:
    better_match = "User B"
    better_score = jaccard_b
    worse_score = jaccard_a
else:
    better_match = "Both users equally"
    better_score = jaccard_a

print(f"\nRECOMMENDATION RESULTS:")
print(f"• User A Similarity: {jaccard_a:.2f}%")
print(f"• User B Similarity: {jaccard_b:.2f}%")
print(f"\n• BEST MATCH FOR FESTIVAL: {better_match} ({better_score:.2f}%)")
print(f"• Difference: {abs(better_score - worse_score):.2f}%")

# Find artists that User B enjoys that User A might like (for cross-recommendation)
user_b_exclusive = user_b.difference(user_a)
artists_in_festival = user_b_exclusive.intersection(festival)

print(f"\n• Artists User B enjoys that are in festival (recommend to User A):")
if artists_in_festival:
    for artist in sorted(artists_in_festival):
        print(f"  → {artist}")

FESTIVAL RECOMMENDATION ANALYSIS

RECOMMENDATION RESULTS:
• User A Similarity: 23.08%
• User B Similarity: 33.33%

• BEST MATCH FOR FESTIVAL: User B (33.33%)
• Difference: 10.26%

• Artists User B enjoys that are in festival (recommend to User A):
  → Eraserheads
  → Indie Artist 4
  → SB19
  → Zild


| Field | Student Entry |
|:------|--------------:|
| **Total Artists in Expanded Festival** | 11 |
| **Length of Intersection (User A & Festival)** | 3 |
| **Length of Union (User A & Festival)** | 13 |
| **User A Jaccard Similarity (%)** | 23.08% |
| **User B Jaccard Similarity (%)** | 33.33% |
| **User A Dealbreaker Artists (Set Output)** | {'Dionela', 'Maki'} |

### **Exercise 3**

In [137]:
# Initialize Base Burst

from collections import deque, Counter

# Initialize deque with strict memory limit of 5
alert_buffer = deque(maxlen=5)

# Simulate burst of incoming alerts
burst = [
    (CONTROL_NUM, "WEATHER"),  # (4, "WEATHER")
    (CONTROL_NUM + 1, "TRAFFIC"),  # (5, "TRAFFIC")
    (CONTROL_NUM + 2, "WEATHER"),  # (6, "WEATHER")
    (CONTROL_NUM + 3, "WEATHER"),  # (7, "WEATHER")
    (CONTROL_NUM + 4, LOCAL_HAZARD),  # (8, "TYPHOON")
    (CONTROL_NUM + 5, "WEATHER"),  # (9, "WEATHER")
    (CONTROL_NUM + 6, "TRAFFIC"),  # (10, "TRAFFIC")
]

print("INITIAL BURST")
for alert in burst:
    print(f"  {alert}")
print(f"Total initial alerts: {len(burst)}")

INITIAL BURST
  (4, 'WEATHER')
  (5, 'TRAFFIC')
  (6, 'WEATHER')
  (7, 'WEATHER')
  (8, 'TYPHOON')
  (9, 'WEATHER')
  (10, 'TRAFFIC')
Total initial alerts: 7


In [138]:
# Data Expansion Phase
# Add exactly (CONTROL_NUM + 2) custom alerts
# CONTROL_NUM = 4, so 4 + 2 = 6 custom alerts needed

print(f"Need to add {CONTROL_NUM + 2} custom alerts (CONTROL_NUM + 2 = 4 + 2 = 6)")

# Find the last timestamp to continue sequentially
last_timestamp = burst[-1][0]

custom_alerts = [
    (last_timestamp + 1, "EARTHQUAKE"),  # (11, "EARTHQUAKE")
    (last_timestamp + 2, "WEATHER"),  # (12, "WEATHER")
    (last_timestamp + 3, "TRAFFIC"),  # (13, "TRAFFIC")
    (last_timestamp + 4, LOCAL_HAZARD),  # (14, "TYPHOON")
    (last_timestamp + 5, "FLOOD"),  # (15, "FLOOD")
    (last_timestamp + 6, "WEATHER"),  # (16, "WEATHER")
]

print(
    f"\nCustom alerts to add (timestamps {last_timestamp + 1} to {last_timestamp + 6}):"
)
for alert in custom_alerts:
    print(f"  {alert}")

# Extend the burst list
burst.extend(custom_alerts)

print("\n=== DATA EXPANSION PHASE ===")
print(f"Added {CONTROL_NUM + 2} custom alerts (timestamps 11-16):")
for alert in custom_alerts:
    print(f"  {alert}")
print(f"\nTotal expanded burst length: {len(burst)}")
print("All alerts in burst (chronological order):")
for i, alert in enumerate(burst, 1):
    print(f"  {i:2d}. {alert}")

Need to add 6 custom alerts (CONTROL_NUM + 2 = 4 + 2 = 6)

Custom alerts to add (timestamps 11 to 16):
  (11, 'EARTHQUAKE')
  (12, 'WEATHER')
  (13, 'TRAFFIC')
  (14, 'TYPHOON')
  (15, 'FLOOD')
  (16, 'WEATHER')

=== DATA EXPANSION PHASE ===
Added 6 custom alerts (timestamps 11-16):
  (11, 'EARTHQUAKE')
  (12, 'WEATHER')
  (13, 'TRAFFIC')
  (14, 'TYPHOON')
  (15, 'FLOOD')
  (16, 'WEATHER')

Total expanded burst length: 13
All alerts in burst (chronological order):
   1. (4, 'WEATHER')
   2. (5, 'TRAFFIC')
   3. (6, 'WEATHER')
   4. (7, 'WEATHER')
   5. (8, 'TYPHOON')
   6. (9, 'WEATHER')
   7. (10, 'TRAFFIC')
   8. (11, 'EARTHQUAKE')
   9. (12, 'WEATHER')
  10. (13, 'TRAFFIC')
  11. (14, 'TYPHOON')
  12. (15, 'FLOOD')
  13. (16, 'WEATHER')


In [139]:
# Initialize tracking variables
rejected_alerts = []
accepted_alerts = []

# Track specific alerts for assessment questions
weather_at_control_plus_3_appended = False  # For timestamp CONTROL_NUM + 3 (7)
local_hazard_appended = False  # For LOCAL_HAZARD (TYPHOON)

print(
    f"Weather at timestamp {CONTROL_NUM + 3} appended: {weather_at_control_plus_3_appended}"
)
print(f"LOCAL_HAZARD ({LOCAL_HAZARD}) appended: {local_hazard_appended}")

Weather at timestamp 7 appended: False
LOCAL_HAZARD (TYPHOON) appended: False


In [140]:
# Process All Alerts Through Sliding Window Filter

# Process each alert through the sliding window filter
for i, (timestamp, category) in enumerate(burst, 1):
    # Step 1: Extract current categories from deque using list comprehension
    current_categories = [alert[1] for alert in alert_buffer]

    # Step 2: Count frequencies using Counter
    category_counter = Counter(current_categories)

    # Step 3: Apply spam rule
    # If category already appears 2 or more times in current buffer, reject
    if category_counter[category] >= 2:
        status = "REJECTED (spam)"
        rejected_alerts.append((timestamp, category))
        decision = "[X] REJECTED"
    else:
        # Accept the alert and append to deque
        alert_buffer.append((timestamp, category))
        status = "ACCEPTED"
        accepted_alerts.append((timestamp, category))
        decision = "[/] ACCEPTED"

        # Track specific alerts for assessment
        if timestamp == CONTROL_NUM + 3:  # Timestamp 7
            weather_at_control_plus_3_appended = True
        if category == LOCAL_HAZARD:  # TYPHOON
            local_hazard_appended = True

    # Print detailed processing log
    print(f"\nAlert {i}: ({timestamp}, '{category}')")
    print(f"  Current buffer categories: {current_categories}")
    print(f"  Category counts: {dict(category_counter)}")
    print(f"  '{category}' count in buffer: {category_counter[category]}")
    print(f"  Decision: {decision}")
    print(f"  Buffer after: {list(alert_buffer)}")

print("\nPROCESS COMPLETE")


Alert 1: (4, 'WEATHER')
  Current buffer categories: []
  Category counts: {}
  'WEATHER' count in buffer: 0
  Decision: [/] ACCEPTED
  Buffer after: [(4, 'WEATHER')]

Alert 2: (5, 'TRAFFIC')
  Current buffer categories: ['WEATHER']
  Category counts: {'WEATHER': 1}
  'TRAFFIC' count in buffer: 0
  Decision: [/] ACCEPTED
  Buffer after: [(4, 'WEATHER'), (5, 'TRAFFIC')]

Alert 3: (6, 'WEATHER')
  Current buffer categories: ['WEATHER', 'TRAFFIC']
  Category counts: {'WEATHER': 1, 'TRAFFIC': 1}
  'WEATHER' count in buffer: 1
  Decision: [/] ACCEPTED
  Buffer after: [(4, 'WEATHER'), (5, 'TRAFFIC'), (6, 'WEATHER')]

Alert 4: (7, 'WEATHER')
  Current buffer categories: ['WEATHER', 'TRAFFIC', 'WEATHER']
  Category counts: {'WEATHER': 2, 'TRAFFIC': 1}
  'WEATHER' count in buffer: 2
  Decision: [X] REJECTED
  Buffer after: [(4, 'WEATHER'), (5, 'TRAFFIC'), (6, 'WEATHER')]

Alert 5: (8, 'TYPHOON')
  Current buffer categories: ['WEATHER', 'TRAFFIC', 'WEATHER']
  Category counts: {'WEATHER': 2, 'T

In [141]:
# Filtering Summary

print(f"Total alerts processed: {len(burst)}")
print(f"Accepted alerts: {len(accepted_alerts)}")
print(f"Rejected alerts: {len(rejected_alerts)}")

print("\n[/] ACCEPTED Alerts (in chronological order):")
for alert in accepted_alerts:
    print(f"  ✓ {alert}")

print("\n[X] REJECTED Alerts (spam filtered):")
for alert in rejected_alerts:
    print(f"  ✗ {alert}")

Total alerts processed: 13
Accepted alerts: 11
Rejected alerts: 2

[/] ACCEPTED Alerts (in chronological order):
  ✓ (4, 'WEATHER')
  ✓ (5, 'TRAFFIC')
  ✓ (6, 'WEATHER')
  ✓ (8, 'TYPHOON')
  ✓ (10, 'TRAFFIC')
  ✓ (11, 'EARTHQUAKE')
  ✓ (12, 'WEATHER')
  ✓ (13, 'TRAFFIC')
  ✓ (14, 'TYPHOON')
  ✓ (15, 'FLOOD')
  ✓ (16, 'WEATHER')

[X] REJECTED Alerts (spam filtered):
  ✗ (7, 'WEATHER')
  ✗ (9, 'WEATHER')


In [142]:
# Final Deque Analysis and Timestamp Summation

# Get final state of deque
final_buffer = list(alert_buffer)
print(f"Final buffer state (maxlen=5): {final_buffer}")
print(f"Final deque length: {len(final_buffer)}")

# Sum all timestamp values in final deque
timestamps_in_buffer = [timestamp for timestamp, _ in final_buffer]
timestamp_sum = sum(timestamps_in_buffer)

print(f"\nTimestamps in final buffer: {timestamps_in_buffer}")
print(f"Sum of timestamps in final deque: {timestamp_sum}")

Final buffer state (maxlen=5): [(12, 'WEATHER'), (13, 'TRAFFIC'), (14, 'TYPHOON'), (15, 'FLOOD'), (16, 'WEATHER')]
Final deque length: 5

Timestamps in final buffer: [12, 13, 14, 15, 16]
Sum of timestamps in final deque: 70


| Field | Student Entry |
|-------|---------------|
| **LOCAL_HAZARD Used** | TYPHOON |
| **Total Burst Size (After adding custom alerts)** | 13 |
| **Did "WEATHER" at Timestamp CONTROL_NUM + 3 append? (Y/N)** | N |
| **Did LOCAL_HAZARD append? (Y/N)** | Y |
| **Final Deque Length** | 5 |
| **Final Sum of Timestamps in Deque** | 70 |
| **Exact Final Deque Output** | [(12, 'WEATHER'), (13, 'TRAFFIC'), (14, 'TYPHOON'), (15, 'FLOOD'), (16, 'WEATHER')] |

### **Summary of Data**

##### 

EXERCISE 1

| Field | Student Entry |
|-------|---------------|
| CONTROL_NUM Used | 4 |
| STUDENT_MAJOR Used | BSECE |
| Expanded Stream Length (After adding data) | 14 |
| Total Applications Processed for TUP Manila | 7 |
| Top Requested Program at TUP Manila | BSECE |
| Exact Frequency of Top Program | 3 |

#####

EXERCISE 2

| Field | Student Entry |
|:------|--------------:|
| **Total Artists in Expanded Festival** | 11 |
| **Length of Intersection (User A & Festival)** | 3 |
| **Length of Union (User A & Festival)** | 13 |
| **User A Jaccard Similarity (%)** | 23.08% |
| **User B Jaccard Similarity (%)** | 33.33% |
| **User A Dealbreaker Artists (Set Output)** | {'Dionela', 'Maki'} |

#####

EXERCISE 3

| Field | Student Entry |
|-------|---------------|
| **LOCAL_HAZARD Used** | TYPHOON |
| **Total Burst Size (After adding custom alerts)** | 13 |
| **Did "WEATHER" at Timestamp CONTROL_NUM + 3 append? (Y/N)** | N |
| **Did LOCAL_HAZARD append? (Y/N)** | Y |
| **Final Deque Length** | 5 |
| **Final Sum of Timestamps in Deque** | 70 |
| **Exact Final Deque Output** | [(12, 'WEATHER'), (13, 'TRAFFIC'), (14, 'TYPHOON'), (15, 'FLOOD'), (16, 'WEATHER')] |

#####

### **Summary of Log**

#####

**EXERCISE 1**

---

**[_*CELL 1*_]**

SEED_NUM: 4  
STUDENT_MAJOR: BSECE  
FAVORITE_ARTIST: RADIOHEAD  
LOCAL_HAZARD: TYPHOON  
CONTROL_NUM: 4

---

**[_*CELL 2*_]**

('TUP Manila', 'BSECE')  
('TUP Taguig', 'BSME_v4')  
('TUP Manila', 'BSME_v4')  
('TUP Manila', 'BSECE')  
('TUP Visayas', 'BSECE')  
('TUP Taguig', 'BSECE')  
('TUP Manila', 'BSECE')  
Initial stream length: 7

---

**[_*CELL 3*_]**

Total stream length: 14  
Custom applications added:  
  1. ('TUP Manila', 'BSECE')  
  2. ('TUP Taguig', 'BSIT_v4')  
  3. ('TUP Visayas', 'RADIOHEAD')  
  4. ('TUP Manila', 'TYPHOON')  
  5. ('TUP Manila', 'BSECE')  
  6. ('TUP Taguig', 'BSECE')  
  7. ('TUP Visayas', 'BSME_v4')

---

**[_*CELL 4*_]**

TUP Manila:  
&emsp;Total applications: 7  
&emsp;Programs: ['BSECE', 'BSME_v4', 'BSECE', 'BSECE', 'BSECE', 'TYPHOON', 'BSECE']  

TUP Taguig:  
&emsp;Total applications: 4  
&emsp;Programs: ['BSME_v4', 'BSECE', 'BSIT_v4', 'BSECE']  

TUP Visayas:  
&emsp;Total applications: 3  
&emsp;Programs: ['BSECE', 'RADIOHEAD', 'BSME_v4']

---

**[_*CELL 5*_]**

TUP Manila Program Frequencies:  
&emsp;BSECE: 5 application(s)  
&emsp;BSME_v4: 1 application(s)  
&emsp;TYPHOON: 1 application(s)  

TUP Taguig Program Frequencies:  
&emsp;BSECE: 2 application(s)  
&emsp;BSME_v4: 1 application(s)  
&emsp;BSIT_v4: 1 application(s)  

TUP Visayas Program Frequencies:  
&emsp;BSECE: 1 application(s)  
&emsp;RADIOHEAD: 1 application(s)  
&emsp;BSME_v4: 1 application(s)  

TUP MANILA TOP PROGRAM ANALYSIS  
&emsp;Top Requested Program: BSECE  
&emsp;Frequency: 5 applications

---

#####

**EXERCISE 2**

---

**[_*CELL 1*_]**

INITIAL FESTIVAL LINEUP  
&emsp;Festival artists: {'Ben&Ben', 'Zild', 'SB19', 'Indie Artist 4', 'RADIOHEAD', 'Bini', 'Eraserheads'}  
&emsp;Total artists: 7  

USER PROFILES  
&emsp;User A: {'Ben&Ben', 'RADIOHEAD', 'Maki', 'Bini', 'Dionela'}  
&emsp;User B: {'Zild', 'SB19', 'Indie Artist 4', 'Parokya ni Edgar', 'Eraserheads'}

---

**[_*CELL 2*_]**

Added 4 new artists to festival:  
  • Munimuni  
  • IVOS  
  • December Avenue  
  • Moira Dela Torre  

Expanded Festival Lineup: {'Ben&Ben', 'IVOS', 'Zild', 'SB19', 'Indie Artist 4', 'RADIOHEAD', 'Moira Dela Torre', 'Bini', 'Munimuni', 'Eraserheads', 'December Avenue'}  
Total artists after expansion: 11

---

**[_*CELL 3*_]**

User A Jaccard Analysis  
&emsp;Intersection (artists in common): {'Bini', 'Ben&Ben', 'RADIOHEAD'}  
&emsp;Size of Intersection: 3  
&emsp;Union (all unique artists): {'Ben&Ben', 'RADIOHEAD', 'Maki', 'Bini', 'December Avenue', 'IVOS', 'Zild', 'SB19', 'Indie Artist 4', 'Moira Dela Torre', 'Munimuni', 'Eraserheads', 'Dionela'}  
&emsp;Size of Union: 13  
&emsp;Jaccard Similarity: 23.08%  

User B Jaccard Analysis  
&emsp;Intersection (artists in common): {'SB19', 'Indie Artist 4', 'Eraserheads', 'Zild'}  
&emsp;Size of Intersection: 4  
&emsp;Union (all unique artists): {'Ben&Ben', 'RADIOHEAD', 'Bini', 'December Avenue', 'IVOS', 'Zild', 'SB19', 'Indie Artist 4', 'Moira Dela Torre', 'Munimuni', 'Parokya ni Edgar', 'Eraserheads'}  
&emsp;Size of Union: 12  
&emsp;Jaccard Similarity: 33.33%

---

**[_*CELL 4*_]**

DEALBREAKERS ANALYSIS  

User A's Listening Profile: {'Ben&Ben', 'RADIOHEAD', 'Maki', 'Bini', 'Dionela'}  
Expanded Festival Lineup: {'Ben&Ben', 'IVOS', 'Zild', 'SB19', 'Indie Artist 4', 'RADIOHEAD', 'Moira Dela Torre', 'Bini', 'Munimuni', 'Eraserheads', 'December Avenue'}  

User A's DEALBREAKERS (artists NOT performing at festival):  
&emsp;✗ Dionela  
&emsp;✗ Maki  

Total dealbreakers: 2  
Dealbreaker percentage: 40.00% of User A's library

---

**[_*CELL 5*_]**

FESTIVAL RECOMMENDATION ANALYSIS  

RECOMMENDATION RESULTS:  
• User A Similarity: 23.08%  
• User B Similarity: 33.33%  

• BEST MATCH FOR FESTIVAL: User B (33.33%)  
• Difference: 10.26%  

• Artists User B enjoys that are in festival (recommend to User A):  
&emsp;→ Eraserheads  
&emsp;→ Indie Artist 4  
&emsp;→ SB19  
&emsp;→ Zild

---

#####

**EXERCISE 3**

---

**[_*CELL 1*_]**

INITIAL BURST  
&emsp;(4, 'WEATHER')  
&emsp;(5, 'TRAFFIC')  
&emsp;(6, 'WEATHER')  
&emsp;(7, 'WEATHER')  
&emsp;(8, 'TYPHOON')  
&emsp;(9, 'WEATHER')  
&emsp;(10, 'TRAFFIC')  
Total initial alerts: 7

---

**[_*CELL 2*_]**

Need to add 6 custom alerts (CONTROL_NUM + 2 = 4 + 2 = 6)  

Custom alerts to add (timestamps 17 to 22):  
&emsp;(17, 'EARTHQUAKE')  
&emsp;(18, 'WEATHER')  
&emsp;(19, 'TRAFFIC')  
&emsp;(20, 'TYPHOON')  
&emsp;(21, 'FLOOD')  
&emsp;(22, 'WEATHER')  

=== DATA EXPANSION PHASE ===  
Added 6 custom alerts (timestamps 11-16):  
&emsp;(17, 'EARTHQUAKE')  
&emsp;(18, 'WEATHER')  
&emsp;(19, 'TRAFFIC')  
&emsp;(20, 'TYPHOON')  
&emsp;(21, 'FLOOD')  
&emsp;(22, 'WEATHER')  

Total expanded burst length: 19  
All alerts in burst (chronological order):  
   1. (4, 'WEATHER')  
   2. (5, 'TRAFFIC')  
   3. (6, 'WEATHER')  
   4. (7, 'WEATHER')  
   5. (8, 'TYPHOON')  
   6. (9, 'WEATHER')  
   7. (10, 'TRAFFIC')  
   8. (11, 'EARTHQUAKE')  
   9. (12, 'WEATHER')  
  10. (13, 'TRAFFIC')  
  11. (14, 'TYPHOON')  
  12. (15, 'FLOOD')  
  13. (16, 'WEATHER')  
  14. (17, 'EARTHQUAKE')  
  15. (18, 'WEATHER')  
  16. (19, 'TRAFFIC')  
  17. (20, 'TYPHOON')  
  18. (21, 'FLOOD')  
  19. (22, 'WEATHER')

---

**[_*CELL 3*_]**

Weather at timestamp 7 appended: False  
LOCAL_HAZARD (TYPHOON) appended: False

---

**[_*CELL 4*_]**

Alert 1: (4, 'WEATHER')  
&emsp;Current buffer categories: []  
&emsp;Category counts: {}  
&emsp;'WEATHER' count in buffer: 0  
&emsp;Decision: [/] ACCEPTED  
&emsp;Buffer after: [(4, 'WEATHER')]  

Alert 2: (5, 'TRAFFIC')  
&emsp;Current buffer categories: ['WEATHER']  
&emsp;Category counts: {'WEATHER': 1}  
&emsp;'TRAFFIC' count in buffer: 0  
&emsp;Decision: [/] ACCEPTED  
&emsp;Buffer after: [(4, 'WEATHER'), (5, 'TRAFFIC')]  

Alert 3: (6, 'WEATHER')  
&emsp;Current buffer categories: ['WEATHER', 'TRAFFIC']  
&emsp;Category counts: {'WEATHER': 1, 'TRAFFIC': 1}  
&emsp;'WEATHER' count in buffer: 1  
&emsp;Decision: [/] ACCEPTED  
&emsp;Buffer after: [(4, 'WEATHER'), (5, 'TRAFFIC'), (6, 'WEATHER')]  

Alert 4: (7, 'WEATHER')  
&emsp;Current buffer categories: ['WEATHER', 'TRAFFIC', 'WEATHER']  
&emsp;Category counts: {'WEATHER': 2, 'TRAFFIC': 1}  
&emsp;'WEATHER' count in buffer: 2  
&emsp;Decision: [X] REJECTED  
&emsp;Buffer after: [(4, 'WEATHER'), (5, 'TRAFFIC'), (6, 'WEATHER')]  

Alert 5: (8, 'TYPHOON')  
&emsp;Current buffer categories: ['WEATHER', 'TRAFFIC', 'WEATHER']  
&emsp;Category counts: {'WEATHER': 2, 'TRAFFIC': 1}  
&emsp;'TYPHOON' count in buffer: 0  
&emsp;Decision: [/] ACCEPTED  
&emsp;Buffer after: [(4, 'WEATHER'), (5, 'TRAFFIC'), (6, 'WEATHER'), (8, 'TYPHOON')]  

Alert 6: (9, 'WEATHER')  
&emsp;Current buffer categories: ['WEATHER', 'TRAFFIC', 'WEATHER', 'TYPHOON']  
&emsp;Category counts: {'WEATHER': 2, 'TRAFFIC': 1, 'TYPHOON': 1}  
&emsp;'WEATHER' count in buffer: 2  
&emsp;Decision: [X] REJECTED  
&emsp;Buffer after: [(4, 'WEATHER'), (5, 'TRAFFIC'), (6, 'WEATHER'), (8, 'TYPHOON')]  

Alert 7: (10, 'TRAFFIC')  
&emsp;Current buffer categories: ['WEATHER', 'TRAFFIC', 'WEATHER', 'TYPHOON']  
&emsp;Category counts: {'WEATHER': 2, 'TRAFFIC': 1, 'TYPHOON': 1}  
&emsp;'TRAFFIC' count in buffer: 1  
&emsp;Decision: [/] ACCEPTED  
&emsp;Buffer after: [(4, 'WEATHER'), (5, 'TRAFFIC'), (6, 'WEATHER'), (8, 'TYPHOON'), (10, 'TRAFFIC')]  

Alert 8: (11, 'EARTHQUAKE')  
&emsp;Current buffer categories: ['WEATHER', 'TRAFFIC', 'WEATHER', 'TYPHOON', 'TRAFFIC']  
&emsp;Category counts: {'WEATHER': 2, 'TRAFFIC': 2, 'TYPHOON': 1}  
&emsp;'EARTHQUAKE' count in buffer: 0  
&emsp;Decision: [/] ACCEPTED  
&emsp;Buffer after: [(5, 'TRAFFIC'), (6, 'WEATHER'), (8, 'TYPHOON'), (10, 'TRAFFIC'), (11, 'EARTHQUAKE')]  

Alert 9: (12, 'WEATHER')  
&emsp;Current buffer categories: ['TRAFFIC', 'WEATHER', 'TYPHOON', 'TRAFFIC', 'EARTHQUAKE']  
&emsp;Category counts: {'TRAFFIC': 2, 'WEATHER': 1, 'TYPHOON': 1, 'EARTHQUAKE': 1}  
&emsp;'WEATHER' count in buffer: 1  
&emsp;Decision: [/] ACCEPTED  
&emsp;Buffer after: [(6, 'WEATHER'), (8, 'TYPHOON'), (10, 'TRAFFIC'), (11, 'EARTHQUAKE'), (12, 'WEATHER')]  

Alert 10: (13, 'TRAFFIC')  
&emsp;Current buffer categories: ['WEATHER', 'TYPHOON', 'TRAFFIC', 'EARTHQUAKE', 'WEATHER']  
&emsp;Category counts: {'WEATHER': 2, 'TYPHOON': 1, 'TRAFFIC': 1, 'EARTHQUAKE': 1}  
&emsp;'TRAFFIC' count in buffer: 1  
&emsp;Decision: [/] ACCEPTED  
&emsp;Buffer after: [(8, 'TYPHOON'), (10, 'TRAFFIC'), (11, 'EARTHQUAKE'), (12, 'WEATHER'), (13, 'TRAFFIC')]  

Alert 11: (14, 'TYPHOON')  
&emsp;Current buffer categories: ['TYPHOON', 'TRAFFIC', 'EARTHQUAKE', 'WEATHER', 'TRAFFIC']  
&emsp;Category counts: {'TYPHOON': 1, 'TRAFFIC': 2, 'EARTHQUAKE': 1, 'WEATHER': 1}  
&emsp;'TYPHOON' count in buffer: 1  
&emsp;Decision: [/] ACCEPTED  
&emsp;Buffer after: [(10, 'TRAFFIC'), (11, 'EARTHQUAKE'), (12, 'WEATHER'), (13, 'TRAFFIC'), (14, 'TYPHOON')]  

Alert 12: (15, 'FLOOD')  
&emsp;Current buffer categories: ['TRAFFIC', 'EARTHQUAKE', 'WEATHER', 'TRAFFIC', 'TYPHOON']  
&emsp;Category counts: {'TRAFFIC': 2, 'EARTHQUAKE': 1, 'WEATHER': 1, 'TYPHOON': 1}  
&emsp;'FLOOD' count in buffer: 0  
&emsp;Decision: [/] ACCEPTED  
&emsp;Buffer after: [(11, 'EARTHQUAKE'), (12, 'WEATHER'), (13, 'TRAFFIC'), (14, 'TYPHOON'), (15, 'FLOOD')]  

Alert 13: (16, 'WEATHER')  
&emsp;Current buffer categories: ['EARTHQUAKE', 'WEATHER', 'TRAFFIC', 'TYPHOON', 'FLOOD']  
&emsp;Category counts: {'EARTHQUAKE': 1, 'WEATHER': 1, 'TRAFFIC': 1, 'TYPHOON': 1, 'FLOOD': 1}  
&emsp;'WEATHER' count in buffer: 1  
&emsp;Decision: [/] ACCEPTED  
&emsp;Buffer after: [(12, 'WEATHER'), (13, 'TRAFFIC'), (14, 'TYPHOON'), (15, 'FLOOD'), (16, 'WEATHER')]  

Alert 14: (17, 'EARTHQUAKE')  
&emsp;Current buffer categories: ['WEATHER', 'TRAFFIC', 'TYPHOON', 'FLOOD', 'WEATHER']  
&emsp;Category counts: {'WEATHER': 2, 'TRAFFIC': 1, 'TYPHOON': 1, 'FLOOD': 1}  
&emsp;'EARTHQUAKE' count in buffer: 0  
&emsp;Decision: [/] ACCEPTED  
&emsp;Buffer after: [(13, 'TRAFFIC'), (14, 'TYPHOON'), (15, 'FLOOD'), (16, 'WEATHER'), (17, 'EARTHQUAKE')]  

Alert 15: (18, 'WEATHER')  
&emsp;Current buffer categories: ['TRAFFIC', 'TYPHOON', 'FLOOD', 'WEATHER', 'EARTHQUAKE']  
&emsp;Category counts: {'TRAFFIC': 1, 'TYPHOON': 1, 'FLOOD': 1, 'WEATHER': 1, 'EARTHQUAKE': 1}  
&emsp;'WEATHER' count in buffer: 1  
&emsp;Decision: [/] ACCEPTED  
&emsp;Buffer after: [(14, 'TYPHOON'), (15, 'FLOOD'), (16, 'WEATHER'), (17, 'EARTHQUAKE'), (18, 'WEATHER')]  

Alert 16: (19, 'TRAFFIC')  
&emsp;Current buffer categories: ['TYPHOON', 'FLOOD', 'WEATHER', 'EARTHQUAKE', 'WEATHER']  
&emsp;Category counts: {'TYPHOON': 1, 'FLOOD': 1, 'WEATHER': 2, 'EARTHQUAKE': 1}  
&emsp;'TRAFFIC' count in buffer: 0  
&emsp;Decision: [/] ACCEPTED  
&emsp;Buffer after: [(15, 'FLOOD'), (16, 'WEATHER'), (17, 'EARTHQUAKE'), (18, 'WEATHER'), (19, 'TRAFFIC')]  

Alert 17: (20, 'TYPHOON')  
&emsp;Current buffer categories: ['FLOOD', 'WEATHER', 'EARTHQUAKE', 'WEATHER', 'TRAFFIC']  
&emsp;Category counts: {'FLOOD': 1, 'WEATHER': 2, 'EARTHQUAKE': 1, 'TRAFFIC': 1}  
&emsp;'TYPHOON' count in buffer: 0  
&emsp;Decision: [/] ACCEPTED  
&emsp;Buffer after: [(16, 'WEATHER'), (17, 'EARTHQUAKE'), (18, 'WEATHER'), (19, 'TRAFFIC'), (20, 'TYPHOON')]  

Alert 18: (21, 'FLOOD')  
&emsp;Current buffer categories: ['WEATHER', 'EARTHQUAKE', 'WEATHER', 'TRAFFIC', 'TYPHOON']  
&emsp;Category counts: {'WEATHER': 2, 'EARTHQUAKE': 1, 'TRAFFIC': 1, 'TYPHOON': 1}  
&emsp;'FLOOD' count in buffer: 0  
&emsp;Decision: [/] ACCEPTED  
&emsp;Buffer after: [(17, 'EARTHQUAKE'), (18, 'WEATHER'), (19, 'TRAFFIC'), (20, 'TYPHOON'), (21, 'FLOOD')]  

Alert 19: (22, 'WEATHER')  
&emsp;Current buffer categories: ['EARTHQUAKE', 'WEATHER', 'TRAFFIC', 'TYPHOON', 'FLOOD']  
&emsp;Category counts: {'EARTHQUAKE': 1, 'WEATHER': 1, 'TRAFFIC': 1, 'TYPHOON': 1, 'FLOOD': 1}  
&emsp;'WEATHER' count in buffer: 1  
&emsp;Decision: [/] ACCEPTED  
&emsp;Buffer after: [(18, 'WEATHER'), (19, 'TRAFFIC'), (20, 'TYPHOON'), (21, 'FLOOD'), (22, 'WEATHER')]  

PROCESS COMPLETE

---

**[_*CELL 5*_]**

Total alerts processed: 13  
Accepted alerts: 11  
Rejected alerts: 2  

[/] ACCEPTED Alerts (in chronological order):  
&emsp;✓ (4, 'WEATHER')  
&emsp;✓ (5, 'TRAFFIC')  
&emsp;✓ (6, 'WEATHER')  
&emsp;✓ (8, 'TYPHOON')  
&emsp;✓ (10, 'TRAFFIC')  
&emsp;✓ (11, 'EARTHQUAKE')  
&emsp;✓ (12, 'WEATHER')  
&emsp;✓ (13, 'TRAFFIC')  
&emsp;✓ (14, 'TYPHOON')  
&emsp;✓ (15, 'FLOOD')  
&emsp;✓ (16, 'WEATHER')  

[X] REJECTED Alerts (spam filtered):  
&emsp;✗ (7, 'WEATHER')  
&emsp;✗ (9, 'WEATHER')

---

**[_*CELL 6*_]**

Final buffer state (maxlen=5): [(12, 'WEATHER'), (13, 'TRAFFIC'), (14, 'TYPHOON'), (15, 'FLOOD'), (16, 'WEATHER')]  
Final deque length: 5  

Timestamps in final buffer: [12, 13, 14, 15, 16]  
Sum of timestamps in final deque: 70

---

#####